### PR TITLE
Bucket publish

### PR DESCRIPTION
**Pull request by mlp6**
Please see individual commit comments for what is going on with this... specifically ``TEMPS`` publish structure renamed ``DATA``, which now includes all temp, valve status and bucket tip data.  Bucket tip is reset after successful publish (based on return status of the command, not an actual handshake confirmation of received data).

I decided not the send ``last_publish_time`` in the publish data to save on bytes, but if that would help on the server end, then we can easily pass that off too.  It could get tricky in the event of a device reset and tracking absolute time, but it is possible.

### CODE REVIEW

File: software/ADPL_electron/ADPL_electron.ino
Comment (suyashkumar) [inline software/ADPL_electron/ADPL_electron.ino:62]: good call factoring `publish_data` out! :D

File: software/ADPL_electron/ADPL_electron.ino
Comment (suyashkumar) [inline software/ADPL_electron/ADPL_electron.ino]: So currently this will publish a bucket tip count every 2.5 minutes no matter what (even if there have been no tips in the last 2.5mins e.g. `bucket.tip_count==0`). I can easily ignore publishes with a bucket-tip count of 0 on the server, but since bucket tips occur in infrequent batches (relative to when temps are sent) we can save 6-7 bytes on most publishes if we so desire. :).

Though it's not perfect, since `sprintf` ignores extra arguments, you can simply switch format strings on the tip count:
```C++
const char* fmt_string = "HXCI:%.1f,HXCO:%.1f,HTR:%.1f,HXHI:%.1f,HXHO:%.1f,V:%d,B:%d";
const char* fmt_string_no_bucket = "HXCI:%.1f,HXCO:%.1f,HTR:%.1f,HXHI:%.1f,HXHO:%.1f,V:%d";

sprintf(data_str,
    (bucket.tip_count > 0) ? fmt_string : fmt_string_no_bucket,
    tempHXCI.temp, tempHXCO.temp, tempHTR.temp, tempHXHI.temp, tempHXHO.temp,
    int(valve.gasOn), int(bucket.tip_count));
```

More succinctly the ternary condition could be `(bucket.tip_count)`, but perhaps it's more readable the other way?

What do you think?

File: software/ADPL_electron/ADPL_electron.ino
Comment (suyashkumar) [reply software/ADPL_electron/ADPL_electron.ino:62]: I also like how this will retry a "failed" publish on the next `loop()` iteration :D

**Review by suyashkumar — COMMENTED** [review]
Looks good! Had a quick comment about only sending bucket tips when the `tip_count > 0` so let me know your thoughts on that!

**Comment by mlp6** [comment]
We could do the same with the valve status and not update that if it hasn't changed since last time.  Also no reason to send the temps if they are not correct (e.g., negative).  When some of these temp probes go offline, we are sending for days or more just lots of nonsense.  

Do you think it is worth putting in this sort of logic for all of the data we are sending?  For the temps, I could modify the ``read()`` method to assign something more useful to that temp value if it is valid (e.g., ``NaN``), and let the downstream functions detect that, or I could let the bad values persist and we just catch them later on.

Thoughts?

**Comment by suyashkumar** [comment]
In short, I think that the logic to have both valve status and bucket tips be optional in the same message begins to get cumbersome in the firmware--since `sprintf` doesn't support named parameters it would be hard to "skip" temperatures or valve status entirely and still have bucket tips if that had changed. So we'd end up having a bunch of different `sprintf` statements that we would switch on in a number of conditions...unless you can think up of a better way to implement that? I think we should try to keep the firmware complexity as low as possible (and as readable as possible) and leave heavy-lifting up to the server because it's easier to debug/push server changes. Same goes for NaNs--i'd rather get the raw data and then decide how to handle it on the server--because it's easier to make/debug server changes than push firmware changes so my philosophy is to keep the firmware as straightforward as possible in certain cases (although this may be some of the Uber firmware-server ideals rubbing off on me).

The server side of things get cumbersome as well when valve status becomes optional--since currently temperatures and valve status are stored together as per what @aforbis-stokes asked for when valve status was added, we would have to build in some additional logic to cache old values or do DB reads of the old values when new messages are received. 

I'd say for now it might be best to move ahead as is to unblock bucket tips feature and ship that feature while we think about how we can best switch out architecture to a "diff" based system that will save us even __more__ data :D. I figured the small optimization above is still readable and succint so was a good one to include in this diff

**Comment by suyashkumar** [comment]
Let me know your thoughts! I think that what you suggest makes a lot of sense and will save us data, but am thinking maybe we move ahead and ship the bucket tips feature first since moving to a diff-based system will require a fair amount of server changes and introduce some additional complexity.

**Comment by mlp6** [comment]
I think that we're in good shape now.  I'll let you merge in once you have the corresponding web server changes made to accommodate the new published data string.

**Review by suyashkumar — APPROVED** [review]

**Comment by suyashkumar** [comment]
Sweet :D. I just verified that nothing will break on the server if these changes are merged as is (the parsing is built to generally accomodate the string we send up--it just won't be accessing the "B" value for now).

Since this change is deployable I'll merge as is and then follow up with the server changes to use the new data being sent up.

**Pull request merged**
